feat: P4ADEV-4082 merge with stashed changes

dan-ntt opened the pull request:
#### Description
<!--- Please always add a PR description as if nobody knows anything about the context these changes come from. -->
<!--- Even if we are all from our internal team, we may not be on the same page. -->
<!--- Write this PR as you were contributing to a public OSS project, where nobody knows you and you have to earn their trust. -->
<!--- This will improve our projects in the long run! Thanks. -->

#### List of Changes
<!--- Describe your changes in detail -->

#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
- Pre-Deploy Test
    - [x] Unit
    - [ ] Integration (Narrow)
- Post-Deploy Test
    - [ ] Isolated Microservice
    - [ ] Broader Integration
    - [ ] Acceptance
    - [ ] Performance & Load

#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] PATCH - Bug fix (backwards compatible bug fixes)
- [x] MINOR - New feature (add functionality in a backwards compatible manner)
- [ ] MAJOR - Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

Commit:
feat: P4ADEV-4082 merge with stashed changes

Commit:
feat: clean version without the new token

Commit:
feat: implement support for limited scope access token creation

antonioT90 commented on src/main/java/it/gov/pagopa/payhub/auth/controller/AuthnControllerImpl.java:
```diff
@@ -34,4 +35,10 @@ public ResponseEntity<Void> logout(String clientId, String token) {
+
+    @Override
+    public ResponseEntity<AccessToken> postLimitedToken(LimitedTokenRequest limitedTokenRequest) {
+        AccessToken accessToken = authnService.postLimitedToken(limitedTokenRequest);
```
let's add a log INFO to trace the request performed adding also some contextual information, like organizationId, app, resource and resourceId

antonioT90 commented on src/main/java/it/gov/pagopa/payhub/auth/controller/AuthzControllerImpl.java:
```diff
@@ -65,7 +65,8 @@ public ResponseEntity<UserInfo> getUserInfoFromMappedExternaUserId(String mapped
         String accessToken = SecurityUtils.getAccessToken();
         UserInfo loggedUser = SecurityUtils.getPrincipal();
-        if(loggedUser.getMappedExternalUserId().equals(mappedExternalUserId)){
+        String loggedExternalUserId = SecurityUtils.getMappedExternalUserId(loggedUser);
```
It would be better to find a way to have a UserInfo with all common properties set, instead of implement a logic to read the property from different models

antonioT90 commented on src/main/java/it/gov/pagopa/payhub/auth/mapper/LimitedScopeTokenMapper.java:
```diff
@@ -0,0 +1,58 @@
+        		.name(baseUserInfo.getName())
+        		.issuer(baseUserInfo.getIssuer())
+        		.organizationAccess(baseUserInfo.getOrganizationAccess())
+        		.organizations(baseUserInfo.getOrganizations())
```
this should be empty instead

antonioT90 commented on src/main/java/it/gov/pagopa/payhub/auth/mapper/LimitedScopeTokenMapper.java:
```diff
@@ -0,0 +1,58 @@
+        		.canManageUsers(baseUserInfo.getCanManageUsers())
+        		.systemUser(baseUserInfo.getSystemUser())
+        		.traceId(baseUserInfo.getTraceId())
+        		.type(baseUserInfo.getType())
```
is this a metadata field used to distinguish between the 2 UserInfo models? If yes, could it be ineherited form the BaseUserInfo?

antonioT90 commented on src/main/java/it/gov/pagopa/payhub/auth/mapper/LimitedScopeTokenMapper.java:
```diff
@@ -0,0 +1,58 @@
+    public LimitedScopeResource mapRequestToLimitedScope(LimitedTokenRequest request) {
+        return LimitedScopeResource.builder()
+        		.app(request.getApp())
+        		.organization(null)
```
here you should copy the selected organization from the organizations on the BaseUserInfo

antonioT90 commented on src/main/java/it/gov/pagopa/payhub/auth/security/JwtAuthenticationFilter.java:
```diff
@@ -52,11 +56,27 @@ protected void doFilterInternal(@Nonnull HttpServletRequest request, @Nonnull Ht
+                    mappedExternalUserId = null;
+                    organizationAccess = null;
+                    organizations = List.of();
+                }
```
this could be reverted if we find a way to have on UserInfo all common fields

antonioT90 commented on src/main/java/it/gov/pagopa/payhub/auth/service/user/IamUserInfoDTO2UserInfoMapper.java:
```diff
@@ -51,7 +66,9 @@ public UserInfo apply(IamUserInfoDTO iamUserInfoDTO, String accessToken) {
-        UserInfo userInfo = UserInfo.builder()
+        BaseUserInfo userInfo = BaseUserInfo.builder()
+                .type(BaseUserInfo.class.getSimpleName())
+                .traceId(iamUserInfoDTO.getTraceId()) // @TODO: identify from where to grab this
```
copy utility method from classification as en example
it.gov.pagopa.pu.classification.util.Utilities#getTraceId

antonioT90 commented on src/main/java/it/gov/pagopa/payhub/auth/service/user/IamUserInfoDTO2UserInfoMapper.java:
```diff
@@ -77,7 +94,9 @@ private UserInfo systemUserMapper(IamUserInfoDTO iamUserInfoDTO, String accessTo
-        UserInfo userInfo = UserInfo.builder()
+        BaseUserInfo userInfo = BaseUserInfo.builder()
+                .type(BaseUserInfo.class.getSimpleName())
+                .traceId(iamUserInfoDTO.getTraceId()) // @TODO: identify from where to grab this
```
use the utility method introduce

antonioT90 commented on src/main/java/it/gov/pagopa/payhub/auth/service/user/IamUserInfoDTO2UserInfoMapper.java:
```diff
@@ -105,6 +124,85 @@ private UserInfo userInfoMapper(IamUserInfoDTO iamUserInfoDTO, String accessToke
+                .organizations(Collections.emptyList()) // @TODO: is this correct?
+                .resource(LimitedScopeResource.builder()
+                        .app(iamUserInfoDTO.getScope())
+                        .resource(null) // @TODO: how to retrieve this info?
```
you should store them on IamUserInfo the entire resource object

Commit:
feat: refactor UserInfo model and update related logic and tests

Commit:
test: add unit tests for limited token scenarios and fix mapping logic

Commit:
test: improve LimitedScopeToken tests and assertions formatting

- Updated test assertions to use specific methods (`assertInstanceOf`, `assertEquals`, `assertNotEquals`) for clarity.
- Rearranged imports in `LimitedTokenServiceImplTest` for better organization.

Commit:
Merge branch 'develop' into P4ADEV-4082-token-limited-scope

# Conflicts:
#	src/main/java/it/gov/pagopa/payhub/auth/utils/Utilities.java

Commit:
feat: update UserInfo mapping and enhance LimitedTokenService tests

- Added `name` and `issuer` fields to UserInfo mapping logic.
- Improved test clarity with structured Given/When/Then comments in `LimitedTokenServiceImplTest`.

Commit:
feat: add organization validation in LimitedTokenService and corresponding tests

- Introduced `validateOrganization` method to ensure user access to requested organization.
- Enhanced test coverage with a new case for unauthorized organization.

Commit:
feat: extend UserInfo mapping and Postman tests for limited token scenarios

- Added `organizations` default field in `UserInfo` mapping logic.
- Enhanced Postman test assertions to validate limited token and UserInfo responses.

antonioT90 commented on src/main/java/it/gov/pagopa/payhub/auth/controller/AuthnControllerImpl.java:
```diff
@@ -34,4 +38,14 @@ public ResponseEntity<Void> logout(String clientId, String token) {
+    public ResponseEntity<AccessToken> postLimitedToken(LimitedTokenRequest limitedTokenRequest) {
+        if (limitedTokenRequest == null) {
+            throw new InvalidScopedAccessTokenRequest("no request body has been provided");
+        }
```
I don't think that these rows are necessary...

antonioT90 commented on src/main/java/it/gov/pagopa/payhub/auth/service/user/IamUserInfoDTO2UserInfoMapper.java:
Introducing a method which accept the UserInfo as object you could avoid many duplication

eg:
`void setCommonSystemUserInfoFields(UserInfo userInfo, IamUserInfoDTO iamUserInfoDTO, String accessToken)`

the same for the other method

Commit:
refactor: streamline UserInfo mapping and enhance limited token handling

- Centralized common `UserInfo` field assignments into helper methods for improved maintainability.
- Updated limited token functionality to support customizable expiration time.
- Adjusted tests to validate changes, including expiration parameter assertions.
- Removed redundant null checks and exception handling for limited token requests.

antonioT90 commented on postman/p4pa-auth-E2E.postman_collection.json:
```diff
@@ -397,7 +398,113 @@
+						"header": [],
+						"body": {
+							"mode": "raw",
+							"raw": "{\n  \"organizationId\": 2,\n  \"app\": \"test\",\n  \"resource\": \"resource\",\n  \"resourceId\": \"resourceId\",\n  \"expireInSeconds\": 14400,\n  \"singleUsage\": false\n}",
```
this collection would be run also on UAT... does this organization exists there? otherwise, you could modify the previous call to getUserInfo to store one of the organization id

antonioT90 reviewed (changes requested):
postman fix

Commit:
feat: enhance Postman tests with dynamic organization selection

- Introduced dynamic retrieval and storage of selected organization details (`organizationId` and `organizationIpaCode`) in collection variables.
- Updated request payloads and assertions to utilize dynamic organization data.
- Added new collection variables for `selectedOrganizationId` and `selectedOrganizationName`.

antonioT90 commented on src/test/java/it/gov/pagopa/payhub/auth/mapper/Client2UserInfoMapperTest.java:
```diff
@@ -15,6 +18,16 @@ class Client2UserInfoMapperTest {
 
+    @BeforeEach
+    void setUp() {
+        MDC.put("traceId", "traceId");
```
you could use `it.gov.pagopa.payhub.auth.utils.UtilitiesTest#setTraceId` and `it.gov.pagopa.payhub.auth.utils.UtilitiesTest#clearTraceIdContext`

Commit:
test: replace MDC usage with UtilitiesTest methods in Client2UserInfoMapperTest

Commit:
test: removed unusued import
